### DISCUSSION TITLE
Merge PDF files with qpdf c++ api

### DISCUSSION THREAD

**Discussion by e-aksenov** [qpdf/qpdf#1191] [Q&A]
Hello,
Is there a sample somewhere how to implement the operation like
> qpdf --empty --pages in1.pdf in2.pdf -- merged.pdf
using C++ API (QPDF , QPDFWriter etc) ? 

I use the following simple approach but not sure if this is the optimal or correct way

void addPages(char const* infile, QPDFPageDocumentHelper& outHelper)
{
    QPDF inpdf;
    inpdf.processFile(infile);
    std::vector<QPDFPageObjectHelper> pages = QPDFPageDocumentHelper(inpdf).getAllPages();
    for (std::vector<QPDFPageObjectHelper>::iterator iter = pages.begin();
        iter != pages.end(); ++iter)
    {
        QPDFPageObjectHelper& page(*iter);
        outHelper.addPage(page, false);
    }
}

int main()
{
        QPDF outpdf;
        outpdf.emptyPDF();
        QPDFPageDocumentHelper outHelper(outpdf);
        addPages("1.pdf", outHelper);
        addPages("2.pdf", outHelper);
        QPDFWriter outpdfw(outpdf, "merged.pdf");
        outpdfw.write();
}

Another issue that comes with this technique is the embedded font subset duplication on each page. Is there a way with qpdf api to merge the font subsets after the merge of pages ?

**Comment by jberkenbilt** [comment]
You can use the QPDFJob API to do anything that you can do with the CLI. The qpdf CLI is a thin wrapper around QPDFJob. I had hoped to create a whole new API for doing page operations, but for various reasons, I'm not optimistic that I will ever actually finish that work.
